core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+  /**
+   * The config to control the instrumentation behavior
+   *
+   * @param thresholdToLogMs If the request completed within the threshold, don't log
+   * @param logRatio         Ratio of requests to log, applied after all funnels/thresholds
+   */
+  case class Config(thresholdToLogMs: Long,
+                    logRatio: Double)
+
+  val ReconfigurableConfigs: Set[String] = Set(
+    KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp,
+    KafkaConfig.LiLongTailProduceRequestLogRatioProp,
+  )
+}
+
+/**
+ * Responsible for logging the instrumentation details.
+ * The config of the logger is associated with DynamicBrokerConfig lifecycle
+ *
+ * @param kafkaConfig For configuring dynamic broker config lifecycle
+ * @param replicaManager Used to obtain the # of partitions in PendingShrinkIsr
+ */
+class ProduceRequestInstrumentationLogger(kafkaConfig: KafkaConfig,
+                                          time: Time,
+                                          rnd: scala.util.Random,
+                                          replicaManager: ReplicaManager) extends Logging with BrokerReconfigurable {
+  override lazy val logger = ProduceRequestInstrumentationLogger.logger
+
+  // This is a cache of the dynamic config.
+  // Should be initialized/reconfigured within the BrokerReconfigurable lifecycle
+  validateReconfiguration(kafkaConfig)
+  kafkaConfig.dynamicConfig.addBrokerReconfigurable(this)
+  var config: Config = extractConfigFromKafkaConfig(kafkaConfig)
+
+  override def reconfigurableConfigs: collection.Set[String] = ProduceRequestInstrumentationLogger.ReconfigurableConfigs
+
+  override def validateReconfiguration(newConfig: KafkaConfig): Unit = {
+    val thresholdMs = newConfig.longTailProduceRequestLogThresholdMs
+    if (thresholdMs < -1) {
+      throw new ConfigException(s"${KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp} should be >= -1")
+    }
+    val ratio = newConfig.longTailProduceRequestLogRatio
+    if (ratio < 0 || ratio > 1) {
+      throw new ConfigException(s"${KafkaConfig.LiLongTailProduceRequestLogRatioProp} should be between 0 to 1")
+    }
+  }
+
+  /**
+   * Invoked when the associated DynamicBrokerConfig is changed
+   */
+  override def reconfigure(oldConfig: KafkaConfig, newConfig: KafkaConfig): Unit = {
+    config = extractConfigFromKafkaConfig(newConfig)
+  }
+
+  def extractConfigFromKafkaConfig(kafkaConfig: KafkaConfig): Config = Config(kafkaConfig.longTailProduceRequestLogThresholdMs,
+                                                                              kafkaConfig.longTailProduceRequestLogRatio)
+
+  /**
+   * Calculate the time in each stage by
+   * 1. Sort stages by timestamp
+   * 2. Make a sliding window of 2 on 1.
+   * 3. Diff the timestamp on the window to get the time spent on the stage of the first element in 2
+   */
+  def toTimeTakenInEachStageMessage(instrumentation: ProduceRequestInstrumentation): String = {
```

Comment:
It would be nice to have a unit test for this function.

Reply:
Added unit tests at
`core/src/test/scala/unit/kafka/server/ProduceRequestInstrumentationTest.scala`